Firestore/core/src/firebase/firestore/local/leveldb_persistence.mm

```diff
@@ -0,0 +1,284 @@
+  if (!status.ok()) {
+    return Status{Error::Internal,
+                  StringFormat("Failed to open LevelDB database at %s",
+                               dir.ToUtf8String())}
+        .CausedBy(ConvertStatus(status));
+  }
+
+  return std::unique_ptr<DB>(database);
+}
+
+// MARK: - LevelDB utilities
+
+LevelDbTransaction* LevelDbPersistence::current_transaction() {
+  HARD_ASSERT(transaction_ != nullptr,
+              "Attempting to access transaction before one has started");
+  return transaction_.get();
+}
+
+util::Status LevelDbPersistence::ClearPersistence(
+    const core::DatabaseInfo& database_info) {
+  Path leveldb_dir = StorageDirectory(database_info, DocumentsDirectory());
+  LOG_DEBUG("Clearing persistence for path: %s", leveldb_dir.ToUtf8String());
+  return util::RecursivelyDelete(leveldb_dir);
+}
+
+int64_t LevelDbPersistence::CalculateByteSize() {
+  int64_t count = 0;
+  auto iter = util::DirectoryIterator::Create(directory_);
+  for (; iter->Valid(); iter->Next()) {
+    int64_t fileSize = util::FileSize(iter->file()).ValueOrDie();
+    count += fileSize;
+  }
+
+  HARD_ASSERT(iter->status().ok(), "Failed to iterate leveldb directory: %s",
+              iter->status().error_message().c_str());
+  HARD_ASSERT(count >= 0 && count <= std::numeric_limits<int64_t>::max(),
+              "Overflowed counting bytes cached");
+  return count;
+}
+
+// MARK: - Persistence
+
+model::ListenSequenceNumber LevelDbPersistence::current_sequence_number() {
+  return reference_delegate_->current_sequence_number();
+}
+
+void LevelDbPersistence::Shutdown() {
+  HARD_ASSERT(started_, "FSTLevelDB shutdown without start!");
+  started_ = false;
+  db_.reset();
+}
+
+LevelDbMutationQueue* LevelDbPersistence::GetMutationQueueForUser(
+    const auth::User& user) {
+  users_.insert(user.uid());
+  current_mutation_queue_ =
+      absl::make_unique<LevelDbMutationQueue>(user, this, serializer_);
+  return current_mutation_queue_.get();
+}
+
+LevelDbQueryCache* LevelDbPersistence::query_cache() {
+  return query_cache_.get();
+}
+
+LevelDbRemoteDocumentCache* LevelDbPersistence::remote_document_cache() {
+  return document_cache_.get();
+}
+
+LevelDbIndexManager* LevelDbPersistence::index_manager() {
+  return index_manager_.get();
+}
+
+LevelDbLruReferenceDelegate* LevelDbPersistence::reference_delegate() {
+  return reference_delegate_.get();
+}
+
+void LevelDbPersistence::RunInternal(absl::string_view label,
+                                     std::function<void()> block) {
+  HARD_ASSERT(transaction_ == nullptr,
+              "Starting a transaction while one is already in progress");
+
+  transaction_ = absl::make_unique<LevelDbTransaction>(db_.get(), label);
+  reference_delegate_->OnTransactionStarted(label);
+
+  block();
+
+  reference_delegate_->OnTransactionCommitted();
+  transaction_->Commit();
+  transaction_.reset();
+}
+
+constexpr char LevelDbPersistence::kReservedPathComponent[];
```

Comment:
Yes. See go/totw/140.

Also note that since I've written this, #3809 has tweaked this formula elsewhere for compatibility with VS 2015. I've adjusted this (and the header) to have the decayed form of `constexpr const char*` instead.